geode/src/main/java/org/apache/calcite/adapter/geode/rel/GeodeFilter.java

```diff
@@ -233,6 +272,85 @@ private String translateOp2(String op, String name, RexLiteral right) {
       }
       return name + " " + op + " " + valueString;
     }
+
+    /**
+     *  Get the field name for the left node to use for IN SET query
+     */
+    private String getLeftNodeFieldName(RexNode left) {
+      switch (left.getKind()) {
+      case INPUT_REF:
+        final RexInputRef left1 = (RexInputRef) left;
+        return fieldNames.get(left1.getIndex());
+      case CAST:
+        // FIXME This will not work in all cases (for example, we ignore string encoding)
+        return getLeftNodeFieldName(((RexCall) left).operands.get(0));
+      case OTHER_FUNCTION:
+        return left.accept(new GeodeRules.RexToGeodeTranslator(this.fieldNames));
+      default:
+        return null;
+      }
+    }
+
+    /**
+     *  Check if we can use IN SET Query clause to improve query performance
+     */
+    private boolean useInSetQueryClause(List<RexNode> disjunctions) {
+      if (disjunctions.size() == 0) {
+        return false;
+      }
+
+      return disjunctions.stream().allMatch(node -> {
+        // IN SET query can only be used for EQUALS
+        if (node.getKind() != SqlKind.EQUALS) {
+          return false;
+        }
+
+        RexCall call = (RexCall) node;
+        final RexNode left = call.operands.get(0);
+        final RexNode right = call.operands.get(1);
+
+        // The right node should always be literal
+        if (right.getKind() != SqlKind.LITERAL) {
+          return false;
+        }
+
+        String name = getLeftNodeFieldName(left);
+        if (name == null) {
+          return false;
+        }
+
+        return true;
+      });
+    }
+
+    /**
+     * Creates OQL IN SET predicate string
+     */
+    private String translateInSet(List<RexNode> disjunctions) {
+      RexNode firstNode = disjunctions.get(0);
```

Comment:
I would add `Preconditions.checkArgument(!disjunctions.isEmpty(), "empty disjunctions")` to fail early